File: JAVA/Weird.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Vector;


public class Weird {

	public static void main(String[] args) {
		try {
			BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
			BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
			int cases = Integer.parseInt(reader.readLine());
			
			while (cases-- > 0) {
				int number = Integer.parseInt(reader.readLine());
				Vector<Integer> setOfDivisors = new Vector<Integer>();
				for (int i = 1; i * i <= number; i++) {
					if (number % i == 0) {
						setOfDivisors.add(i);
						if (i != number / i) setOfDivisors.add(number/i);
					}
				}
				setOfDivisors.remove((Object)number);
				quickSortOrderByAsce(0, setOfDivisors.size() - 1, setOfDivisors);
				int sumOfDivisor = 0;
				for (int i = 0; i < setOfDivisors.size(); i++) sumOfDivisor += setOfDivisors.get(i); 
				if (sumOfDivisor > number && !checkWeirdNumber(setOfDivisors, number)) writer.append("weird");
				else writer.append("not weird");
				writer.append("\n");
			}
			writer.flush();
			writer.close(); reader.close();
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
	
	public static boolean checkWeirdNumber(Vector<Integer> array, int number) {
		int sumOfDivisor = 0;
		for (int i = 0; i < array.size(); i++) {
			sumOfDivisor += array.get(i);
			if (sumOfDivisor == number) return true;
			else if (sumOfDivisor < number) return checkPseudoPerfectNumber(array, number, sumOfDivisor, i);
			sumOfDivisor -= array.get(i);
		}
		return false;
	}
	
	public static boolean checkPseudoPerfectNumber(Vector<Integer> array, int number, int sumOfDivisor, int index) {
		for (int i = index + 1; i < array.size(); i++) {
			sumOfDivisor += array.get(i);
			if (sumOfDivisor == number) return true;
			else if (sumOfDivisor < number) return checkPseudoPerfectNumber(array, number, sumOfDivisor, i);
			sumOfDivisor -= array.get(i);
		}
		return false;
	}
	
	public static void quickSortOrderByAsce(int left, int right, Vector<Integer> array) {
		int h_left = left;
		int h_right = right;
		int pivot = array.get(left);
		while (left < right) {
			while (array.get(right) < pivot && left < right) right--;
			if (left != right) array.set(left, array.get(right)); 
			while (array.get(left) > pivot && left < right) left++;
			if (left != right) {
				array.set(right, array.get(left)); right--;
			}
		}
		array.set(left, pivot);
		pivot = left;
		left = h_left;
		right = h_right;
		if (left < pivot) quickSortOrderByAsce(left, pivot - 1, array);
		if (pivot < right) quickSortOrderByAsce(pivot + 1, right, array);
	}

}

File: CPP/Weird.cpp
#include <cstdio>
#include <vector>
using namespace std;

void quickSortOrderByAsce(int left, int right, vector<int>& array) {
	int h_left = left, h_right = right, pivot = array.at(left);
	while (left < right) {
		while (pivot > array.at(right) && left < right) right--;
		if (left != right) array.at(left) = array.at(right);
		while (pivot < array.at(left) && left < right) left++;
		if (left != right) { array.at(right) = array.at(left); right--;}
	}
	array.at(left) = pivot;
	pivot = left;
	left = h_left; right = h_right;
	if (left < pivot) quickSortOrderByAsce(left, pivot - 1, array);
	if (pivot < right) quickSortOrderByAsce(pivot + 1, right, array);
}

bool checkPseudoPecfectNumber(vector<int> array, int number, int sum, int index) {
	for (int i = index; i < (int)array.size(); i++) {
		sum += array.at(i);
		if (sum == number) return true;
		else if (sum < number) return checkPseudoPecfectNumber(array, number, sum, i + 1);
		sum -= array.at(i);
	}
	return false;
}

int main() {
	int cases;
	scanf("%d", &cases);

	while (cases--) {
		int number, sumOfDivisor = 0; scanf("%d", &number);
		vector<int> setOfDivisor;
		for (int i = 1; i * i <= number; i++) {
			if (number % i == 0) {
				setOfDivisor.push_back(i);
				if (i != 1 && i != number / i) setOfDivisor.push_back(number/i);
			}
		}
		quickSortOrderByAsce(0, (int)setOfDivisor.size() - 1, setOfDivisor);
		for (int i = 0; i < (int)setOfDivisor.size(); i++) sumOfDivisor += setOfDivisor.at(i);
		if (sumOfDivisor < number || checkPseudoPecfectNumber(setOfDivisor, number, 0, 0)) fputs("not weird\n", stdout);
		else fputs("weird\n", stdout);
	}
}



